### PR TITLE
[Filesystem] Add the Path class

### LINKED ISSUE #26523: [Filesystem] Add $fs->join([$pathSegment])
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | latest

When working with filesystem it's always very cumbersome to create path from different segments. For example, create full path from base directory, subdirectory and filename. Especially if you need to compare paths later. 

What I propose is to add `$filesystem->join()` method. It's purpose would be to merge different path segment with path separator in a predictable way. 

```php
$filesystem->join(['some', 'path', 'segment']); // 'some/path/segment'

$filesystem->join(['/some/path/', '/subdir/', 'filename.tmp']); // '/some/path/subdir/filename.tmp'
// notice how it behaves better than implode('/', $array) in regards to doubleslashes

// following OS depended behavior is up to discussion
$filesystem->join(['c:\', '/subdir/', 'filename.tmp']); // 'c:/subdir/filename.tmp'
// that this way one can be sure that slash is separator and it works fine due to PHP transforming it on the fly when accessing the file

// also this could be useful
$filesystem->toOsSpecific('some/path\to\some/folder'); 
// 'some\path\to\some\folder' on windows
// 'some/path/to/some/folder' on linux
$filesystem->toPathWithSlashes('some\path/to\some\folder');
// always 'some/path/to/some/folder'
```

What do you think? I'd be happy to send a PR after it's disussed.

### PR DESCRIPTION
**Pull request by theofidry**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #26523
| License       | MIT
| Doc PR        | TODO

Re-opening #30969 as requested by \@fabpot.

This PR proposes a low-level, performant and pure API to manipulate file paths.

Two methods `Path::isAbsolute()` and `Path::isRelative()` are similar to `Filesystem::isAbsolutePath()` and `Filesystem::makePathRelative()`. They are however different as the methods from `Path` do not check the existence of the path which is very helpful in some scenarios (for example when manipulating virtual paths such as "regular" paths of a PHAR).

The original code comes from [webmozart/path-util](https://github.com/webmozart/path-util) from which the path part was extracted. The library is very stable, well tested and documented. It has been heavily used in [Puli](https://github.com/puli), which although is no longer active has been a good performance test and is used in several other popular projects such as:

- vimeo/psalm
- drush/drush
- infection/infection
- humbug/box
- humbug/php-scoper
- phpbench/phpbench

It however suffers from lack of maintener from @webmozart which no longer has time to maintain it.

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Path.php [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+
+    /**
+     * Canonicalizes the given path.
+     *
+     * During normalization, all slashes are replaced by forward slashes ("/").
+     * Furthermore, all "." and ".." segments are removed as far as possible.
+     * ".." segments at the beginning of relative paths are not removed.
+     *
+     * ```php
+     * echo Path::canonicalize("\webmozart\puli\..\css\style.css");
```
Comment (OskarStark) [inline src/Symfony/Component/Filesystem/Path.php]: While I would agree keeping this as credits, should we change "webmozart" to "symfony" in examples?

Comment (theofidry) [reply src/Symfony/Component/Filesystem/Path.php]: Changed it for `Path` (left it for the tests)

Comment (OskarStark) [reply src/Symfony/Component/Filesystem/Path.php]: Thank you

File: src/Symfony/Component/Filesystem/Exception/RuntimeArgumentException.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Filesystem\Exception;
+
+/**
+ * @author Théo Fidry <theo.fidry@gmail.com>
+ */
+class RuntimeArgumentException extends \RuntimeException implements ExceptionInterface
```
Comment (chalasr) [inline src/Symfony/Component/Filesystem/Exception/RuntimeArgumentException.php]: This exception does not seem to be used, do I miss something?

Comment (theofidry) [reply src/Symfony/Component/Filesystem/Exception/RuntimeArgumentException.php]: Hm I think it was meant to replace the usage of `\RuntimeException` in `Path`. I renamed it and replace the usage

**Review by chalasr — APPROVED** [review]
Looks good to me. 
Still pretty sure there are some places in core that could leverage this. That might be for another PR though.

**Comment by theofidry** [comment]
> Still pretty sure there are some places in core that could leverage this. That might be for another PR though.

I am not familiar enough with the code-base to spot those. But as a rule of thumb, I would keep the current usages of `Filesystem::isAbsolutePath()` as it is a stronger check than `Path::isAbsolute()`

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Filesystem/Path.php [hunk 3]
```diff
@@ -0,0 +1,821 @@
...
+        if (isset(self::$buffer[$path])) {
+            return self::$buffer[$path];
+        }
+
+        // Replace "~" with user's home directory.
+        if ('~' === $path[0]) {
+            $path = self::getHomeDirectory().mb_substr($path, 1);
+        }
+
+        $path = str_replace('\\', '/', $path);
```
Comment (YaFou) [inline src/Symfony/Component/Filesystem/Path.php]: Why not use the normalize method?

**Comment by OskarStark** [comment]
Can you please target 🎯 6.0, as we are not allowed to introduce experimental classes in an LTS version.

**Comment by Nyholm** [comment]
> Can you please target 🎯 6.0, as we are not allowed to introduce experimental classes in an LTS version.

@OskarStark is correct. We cannot have experimental code in LTS. However, we can make this new class non-experimental. It is a well used class outside of Symfony as the PR description tells. 

@theofidry, Do what you prefer.

**Comment by chalasr** [comment]
Since this has already proven useful and it’s not controversial, I vote for 5.4 with no experimental flag.

File: src/Symfony/Component/Filesystem/Path.php (hunk 3 above)
Comment (theofidry) [reply src/Symfony/Component/Filesystem/Path.php]: No reason, changed it

**Comment by chalasr** [comment]
Please rebase :) We cannot merge PRs that contain merge commits

**Comment by theofidry** [comment]
>Please rebase :) We cannot merge PRs that contain merge commits

Is there a technical reason for that choice or is it just the process in place? Seems weird to have to rebase which requires  a force-push from the user and completely change the commits history (which messes up the GitHub discussions)

**Comment by chalasr** [comment]
Yes, IIRC it's a limitation from the internal tool we use for merging, it doesn't play well with merge commits

File: src/Symfony/Component/Filesystem/Path.php [hunk 4]
```diff
@@ -0,0 +1,821 @@
...
+ *
+ * The methods in this class are able to deal with both UNIX and Windows paths
+ * with both forward and backward slashes. All methods return normalized parts
+ * containing only forward slashes and no excess "." and ".." segments.
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Thomas Schulz <mail@king2500.net>
+ * @author Théo Fidry <theo.fidry@gmail.com>
+ *
+ * @experimental in 5.4
```
Comment (chalasr) [inline src/Symfony/Component/Filesystem/Path.php:28]: to be removed

File: src/Symfony/Component/Filesystem/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+5.4.0
+-----
+
+* Add `Path`
```
Comment (OskarStark) [inline src/Symfony/Component/Filesystem/CHANGELOG.md]:
```suggestion
* Add `Path` class
```
?

**Comment by webmozart** [comment]
Just adding my two cents: This PR (including its predecessor #30969) has been open for over two years (!). It's not very motivating for a contributor to have their work pending for so long, and then being asked to make small changes, rebases etc. every few months.

@theofidry rebased this 19 days ago. I think it would be a great appreciation of his work and endurance to merge this, rather than waiting again, asking for another rebase, adding more nitpick comments etc.

File: src/Symfony/Component/Filesystem/Path.php (hunk 4 above)
Comment (chalasr) [inline src/Symfony/Component/Filesystem/Path.php]:
```suggestion
```

**Comment by chalasr** [comment]
Thank you @theofidry.

**Pull request merged**